File: src/com/kedzie/vbox/api/IMedium.java
package com.kedzie.vbox.api;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import android.os.Parcel;
import android.os.Parcelable;

import com.kedzie.vbox.api.jaxb.DeviceType;
import com.kedzie.vbox.api.jaxb.MediumState;
import com.kedzie.vbox.api.jaxb.MediumType;
import com.kedzie.vbox.soap.Asyncronous;
import com.kedzie.vbox.soap.KSOAP;
import com.kedzie.vbox.soap.VBoxSvc;

@KSOAP
public interface IMedium extends IManagedObjectRef, Parcelable {
	public static final String BUNDLE = "medium";
	static final ClassLoader LOADER = IMedium.class.getClassLoader();
	
	public static final Parcelable.Creator<IMedium> CREATOR = new Parcelable.Creator<IMedium>() {
		public IMedium createFromParcel(Parcel in) {
			VBoxSvc vmgr =  in.readParcelable(LOADER);
			String id = in.readString();
			Map<String, Object> cache = new HashMap<String, Object>();
			in.readMap(cache, LOADER);
			return (IMedium) vmgr.getProxy(IMedium.class, id, cache); 
		}
		public IMedium[] newArray(int size) {  
			return new IMedium[size]; 
		}
	};
	
     @KSOAP(cacheable=true) public String getId();
     
     @KSOAP(cacheable=true) public String getDescription();
     @Asyncronous public void setDescription(@KSOAP("description")String description);
     
     @KSOAP(cacheable=true) public MediumState getState();
     
     @KSOAP(cacheable=true) public Integer getVariant();

     @KSOAP(cacheable=true) public String getLocation();
     @Asyncronous public void setLocation(@KSOAP("location")String location);
     
     @KSOAP(cacheable=true) public String getName();
     
     @KSOAP(cacheable=true) public DeviceType getDeviceType();
     
     @KSOAP(cacheable=true) public Boolean getHostDrive();

     @KSOAP(cacheable=true) public Long getSize();

     @KSOAP(cacheable=true) public String getFormat();
     
     @KSOAP(cacheable=true) public IMediumFormat getMediumFormat();
     
     @KSOAP(cacheable=true) public MediumType getType();
     @Asyncronous public void setType(@KSOAP("type") MediumType type);
     
     @KSOAP(cacheable=true) public MediumType[]  getAllowedTypes();
     
     @KSOAP(cacheable=true) public IMedium getParent();
     
     @KSOAP(cacheable=true) public List<IMedium> getChildren();
     
     @KSOAP(cacheable=true) public IMedium getBase();

     @KSOAP(cacheable=true) public Boolean getReadOnly();

     @KSOAP(cacheable=true) public Long getLogicalSize();

     @KSOAP(cacheable=true) public Boolean getAutoReset();
     @Asyncronous public void setAutoReset(@KSOAP("autoReset") boolean autoReset);
     
     @KSOAP(cacheable=true) public String getLastAccessError();

     @KSOAP(cacheable=true) public List<String> getMachineIds();
     
     public void setIds(@KSOAP("setImageId") boolean setImageId, @KSOAP("imageId") String imageId, @KSOAP("setParentId") boolean setParentId, @KSOAP("parentId") String parentId);
     
     public MediumState refreshState();

     @KSOAP(cacheable=true) public List<String> getSnapshotIds();

     public MediumState lockRead();
     public MediumState unlockRead();

     public MediumState lockWrite();
     public MediumState unlockWrite();

     public void close();

     @KSOAP(cacheable=true) String getProperty(@KSOAP("key") String key);
 	@Asyncronous public void setProperty(@KSOAP("key") String key, @KSOAP("value") String value);
 	@KSOAP(cacheable=true) public Map<String, List<String>> getProperties(@KSOAP("names") String names);
 	@Asyncronous public void setProperties(@KSOAP("names") List<String> names, @KSOAP("values") List<String> values);

 	public IProgress createBaseStorage(@KSOAP(type="long", value="logicalSize") long logicalSize, @KSOAP(type="unsignedInt", value="variant") int variant);
 	
 	public IProgress createDiffStorage(@KSOAP(type="long", value="logicalSize") long logicalSize, @KSOAP(type="unsignedInt", value="variant") int variant);

 	public IProgress deleteStorage();

 	public IProgress mergeTo(@KSOAP("target") IMedium target);

 	public IProgress cloneTo(@KSOAP("target") IMedium target, @KSOAP(type="unsignedInt", value="variant")int variant, @KSOAP("parent") IMedium parent);

 	public IProgress cloneToBase(@KSOAP("target") IMedium target, @KSOAP(type="unsignedInt", value="variant")int variant);

 	public IProgress compact();

 	public IProgress resize(@KSOAP(type="long", value="logicalSize") long logicalSize);
 	
 	public IProgress reset();
}

File: vbox-app/src/main/java/com/kedzie/vbox/api/IMedium.kt
package com.kedzie.vbox.api

import android.os.Parcelable
import com.kedzie.vbox.api.jaxb.DeviceType
import com.kedzie.vbox.api.jaxb.MediumState
import com.kedzie.vbox.api.jaxb.MediumType
import com.kedzie.vbox.app.Utils
import com.kedzie.vbox.soap.Cacheable
import com.kedzie.vbox.soap.Ksoap
import com.kedzie.vbox.soap.KsoapProxy
import java.io.IOException
import java.util.*

@KsoapProxy
@Ksoap
interface IMedium : IManagedObjectRef, Parcelable {
    companion object {
        const val BUNDLE = "medium"
    }

    @Cacheable("Id")
	suspend fun getId(): String

    @Cacheable("Description")
    suspend fun getDescription(): String
    suspend fun setDescription(@Cacheable("Description") description: String)

    @Cacheable("State")
	suspend fun getState(): MediumState

    @Cacheable("Variant")
	suspend fun getVariant(): Int

    @Cacheable("Location")
	suspend fun getLocation(): String;
    suspend fun setLocation(@Cacheable("Location") location: String)

    @Cacheable("Name")
	suspend fun getName(): String

    @Cacheable("DeviceType")
	suspend fun getDeviceType(): DeviceType

    @Cacheable("HostDrive")
	suspend fun getHostDrive(): Boolean

    @Cacheable("Size")
	suspend fun getSize(): Long

    @Cacheable("Format")
	suspend fun getFormat(): String

    @Cacheable("MediumFormat")
	suspend fun getMediumFormat(): IMediumFormat

    @Cacheable("Type")
	suspend fun getType(): MediumType
    suspend fun setType(@Cacheable("Type") type: MediumType)

    @Cacheable
    suspend fun getAllowedTypes(): Array<MediumType>

    @Cacheable("Parent")
	suspend fun getParent(): IMedium

    @Cacheable("Children")
	suspend fun getChildren(): List<IMedium>

    @Cacheable("Base")
	suspend fun getBase(): IMedium

    @Cacheable("ReadOnly")
	suspend fun getReadOnly(): Boolean

    @Cacheable("LogicalSize")
	suspend fun getLogicalSize(): Long

    @Cacheable("AutoReset")
	suspend fun getAutoReset(): Boolean;
    suspend fun setAutoReset(@Cacheable("AutoReset") autoReset: Boolean)

    @Cacheable("LastAccessError")
	suspend fun getLastAccessError(): String;

    @Cacheable("MachineIds")
	suspend fun getMachineIds(): List<String>

    suspend fun setIds(setImageId: Boolean, imageId: String, setParentId: Boolean, parentId: String)

    suspend fun refreshState(): MediumState

    @Cacheable("SnapshotIds")
    suspend fun getSnapshotIds(): List<String>

    suspend fun lockRead(): MediumState
    suspend fun unlockRead(): MediumState

    suspend fun lockWrite(): MediumState
    suspend fun unlockWrite(): MediumState

    suspend fun close()

    suspend fun getProperty(key: String): String

    suspend fun setProperty(key: String, value: String);

    suspend fun getProperties(names: String) : Map<String, List<String>>

    suspend fun setProperties(names: List<String>, values: List<String>)

    suspend fun createBaseStorage(@Ksoap(type="long") logicalSize: Long, @Ksoap(type="unsignedInt") variant: Int): IProgress

    suspend fun createDiffStorage(@Ksoap(type="long") logicalSize: Long, @Ksoap(type="unsignedInt") variant: Int): IProgress

    suspend fun deleteStorage():IProgress

    suspend fun mergeTo(target: IMedium): IProgress

    suspend fun cloneTo(target: IMedium, @Ksoap(type="unsignedInt") variant: Int, parent: IMedium): IProgress

    suspend fun cloneToBase(target: IMedium, @Ksoap(type="unsignedInt") variant: Int): IProgress

    suspend fun compact(): IProgress

    suspend fun resize(@Ksoap(type="long") logicalSize: Long): IProgress

    suspend fun reset(): IProgress
}

/**
 * Load medium properties
 * @param names  Property names to load, or empty for all
 * @return    properties
 */
suspend fun IMedium.getProperties(vararg names: String): Properties {
    val nameString = StringBuffer()
    for (name in names)
        Utils.appendWithComma(nameString, name)

    val map = getProperties(nameString.toString())
    val returnNames = map.get("returnNames")
    val values = map.get("returnval")
    val properties = Properties()
    for (i in returnNames!!.indices)
        properties[returnNames!![i]] = values!![i]
    return properties
}